modules/videoio/src/cap_v4l.cpp

```diff
@@ -273,8 +272,8 @@ struct CvCaptureCAM_V4L CV_FINAL : public CvCapture
     int bufferIndex;
     int FirstCapture;
     String deviceName;
+    mutable cv::Mutex mutex;
```

Comment:
I do not think that locking should be introduced at a per-backend level. If desired it should be done at the VideoCapture toplevel.

Reply:
Agreed.
Extra synchronization should be avoided on backend level if it is not necessary (async callbacks).

Reply:
So, what is your purpose? Removing mutex/CriticalSection from all backend level and creating it at the VideoCapture level? Yes, this is good idea, but you don't have it now. For now, I want to get capture on the linux (v4l) same as on the windows (msmf).

Reply:
you should not assume the backends to be threadsafe. Rather the msmf is threadsafe by coincidence as it needs the locks due to the internal async callbacks.
I would rather suggest introducing locks on application level so singlethreaded apps do not have to pay the penalty.

Locking on VideoCapture level should be a separate PR and probably would incur a separate discussion.

Reply:
"threadsafe by coincidence" - try to remove synchronization from cap_dshow, cap_msmf, cap_ffmpeg_impl (it have it's own wrapper :) ), cap_gstreamer, cap_openni2, cap_ffmpeg. Only for VideoCapture_WinRT you move mutex to VideoCapture. Why? Maybe the problem is that a backend synchronization is a backend specific? 

"you should not assume the backends to be threadsafe." Why? I configure and debug the capture on one platform and I have to do the same on the other, isn't it? Is it a good idea to force OpenCV users to write platform-specific wrappers for OpenCV objects?

"singlethreaded apps do not have to pay the penalty" I use cv::Mutex, if you say that it is slow than I could use something else or wrap it to `#ifdef USE_THREADSAFE_OPENCV_V4L`. Is the mutex really slower than converting to RGB or even memcopy of an image?

Reply:
> try to remove synchronization from cap_dshow (..)

I was talking about VideoCapture level thread-safety there :confused: 

>  Is the mutex really slower than converting to RGB (..)

I agree that the overhead is likely neglectible. However I think if we make VideoCapture threadsafe, same reasoning also applies to the [Algorithm](https://docs.opencv.org/master/d3/d46/classcv_1_1Algorithm.html) classes.

Reply:
Multi-threaded code is usually more complicated. So this is the code maintenance question at first. Also it is hard to test synchronization or reproduce issues.

If synchronization is not required to use library from single-threaded app (or can be successfully handled on caller level), then please avoid it here.

Reply:
I think, a thread-safety is a backend specific task. Let's holy war begin! :) Really, users don't care about synchronization destructor of backend object and getting parameter from a camera. Also, for some backends, getting parameters are thread-safe permanently and for other one you can't get anything while someone is already reading from it.

> same reasoning also applies to the Algorithm classes

Why not? Let user make a choice:  singlethread mode or thread-safe. For  singlethread you could make  cv::Mutex more simple. Or make cv::Mutex faster (depending on platform)